node/src/components/linear_chain.rs

```diff
@@ -416,26 +517,35 @@ where
                     }
                 }
             }
-            Event::IsBonded(Some(block), fs, true) => {
+            Event::IsBonded(Some(mut block), fs, true) => {
                 // Known block and signature from a bonded validator.
-                self.add_pending_finality_signature(*fs);
-                let old_count = block.proofs().len();
-                let (block, mut effects) =
-                    self.collect_pending_finality_signatures(*block, effect_builder);
-                if block.proofs().len() > old_count {
-                    let block = Box::new(block);
+                // Check if we had already seen this signature before.
+                let signature_known = block
+                    .proofs()
+                    .get(&fs.public_key)
+                    .iter()
+                    .any(|sig| *sig == &fs.signature);
+                // If new, gossip and store.
+                if !signature_known {
```

Comment:
It would be easier to read if the positive case came first:
```rust
if signature_known {
    Effects::new()
} else {
```

Reply:
Done in [`6343901` (#773)](https://github.com/CasperLabs/casper-node/pull/773/commits/6343901b9b4b5c4e24afb6226245306bdd76b8c0)